src/coreclr/jit/llvm.cpp

```diff
@@ -1589,7 +1589,7 @@ void Llvm::buildStoreInd(GenTreeStoreInd* storeIndOp)
 {
     Value* address = getGenTreeValue(storeIndOp->Addr());
     Value* toStore = getGenTreeValue(storeIndOp->Data());
-    if (toStore->getType()->isPointerTy() && (storeIndOp->gtFlags & GTF_IND_TGT_NOT_HEAP) == 0)
+    if (toStore->getType()->isPointerTy() && (storeIndOp->gtFlags & GTF_IND_TGT_NOT_HEAP) == 0 && address->getType()->isPointerTy())
```

Comment:
I don't think this is correct. The address valid can be an integer here, that points to the heap, to a pinned object.

(Side note: we should `consumeValue` it as a pointer to the type we're going store instead of manually `castIfNecessary`ing)

The logic required to know whether we need the barrier should be identical to this: https://github.com/dotnet/runtime/blob/10ac54d21b60263a94596cba441bdd68f370d480/src/coreclr/jit/gcinfo.cpp#L224-L272

(`WBF_BarrierUnchecked` is `RhpAssignRef`, `WBF_BarrierChecked` and `WBF_Unknown` are `RhpCheckedAssignRef`)

It would be nice to reuse it here, if it is not possible, the simplified (inefficient) version is that we need a write barrier for `storeIndOp->TypeIs(TYP_REF)`, and it should be checked.

Reply:
I can change this to `consumeValue`,  but I'll still need to cast to `i8*` as that will be the types for the args in `RhpAssignRef` and `RhpCheckedAssignRef`   LLVM is going to opaque pointers which might remove this (i've not read up on that).